implement user management feature

kevin5603 opened the pull request:
# default user

email: admin@a.com, password: 111111
email: manager@a.com, password: 111111

kevin5603 committed:
ignore idea settings

kevin5603 committed:
Use devise & create table

kevin5603 committed:
fix webpack issue & minor

kevin5603 committed:
custom devise views

kevin5603 committed:
Add bootstrap & add user column

kevin5603 committed:
Add Devise confirmed module

kevin5603 committed:
minor

kevin5603 committed:
Add cancancan for authorization

kevin5603 committed:
Add sidekiq library for registration notification

kevin5603 committed:
fix bug & user CRUD

kevin5603 committed:
erb to haml & refactor

kevin5603 committed:
minor

Aronyu127 commented on app/controllers/users_controller.rb:
```diff
@@ -0,0 +1,39 @@
+class UsersController < ApplicationController
+  before_action :authenticate_user!
+  before_action :set_user, only: %i[show edit update destroy]
```
可以考慮用  `load_and_authorize_resource`

https://github.com/CanCanCommunity/cancancan/blob/develop/docs/controller_helpers.md#authorize_resource-load_resource-load_and_authorize_resource

Aronyu127 commented on app/helpers/users_helper.rb:
```diff
@@ -0,0 +1,14 @@
+module UsersHelper
```
改 `load_and_authorize_resource` 之後這個檔案應該就不需要了

Aronyu127 commented on app/mailers/admin_mailer.rb:
```diff
@@ -0,0 +1,11 @@
+  default from: "no-reply@tao.user.management.com"
+
+  def registration_notification(admin_email_list, register_email, first_name, last_name)
+    @first_name = first_name
```
實務上的做法通常是傳id 進來
這邊init user 後 再去 render 想顯示的資料，不然今天要改 顯示電話 整個 function 都要一起改

Aronyu127 commented on app/models/user.rb:
```diff
@@ -0,0 +1,20 @@
+  private
+
+  def assign_default_role
+    default_role = Role.find_or_create_by(name: 'regular_user')
```
regular_user 是 default role 這件事情可以寫在 Role model裡面， 這邊只要拿就好

Aronyu127 commented on app/sidekiq/registration_notification_job.rb:
```diff
@@ -0,0 +1,8 @@
+  include Sidekiq::Job
+
+  def perform(register_email, first_name, last_name)
+    @admin_email_list = User.includes(:roles).where(roles: { name: 'admin' }).map(&:email)
```
`.map(&:email)` 可以試試 `.pluck(:email)`

Aronyu127 commented on app/views/devise/shared/_error_messages.html.haml:
```diff
@@ -0,0 +1,9 @@
+- if resource.errors.any?
```
這個顯示應該可以 form 都共用，不局限於 devise

Aronyu127 commented on app/views/users/edit.html.haml:
```diff
@@ -0,0 +1,15 @@
+    %label Roles
+    %br/
+    - @roles.each do |role|
+      = check_box_tag "user[role_ids][]", value=role.id, @user.roles.map(&:name).include?(role.name)
```
用 ID 比對應該比較精準, 畢竟name 目前沒有限制不能重複

Aronyu127 commented on app/views/users/index.html.haml:
```diff
@@ -0,0 +1,31 @@
+          %td= user.job_title
+          %td= user.roles.map(&:name).join(", ")
+          %td= user.confirmed_at ? "Enabled" : "Disabled"
+          - if current_user.roles.map(&:name).include?("admin")
```
前端的顯示控制可以參考 https://github.com/CanCanCommunity/cancancan/blob/develop/docs/check_abilities_mistakes.md
來處理

Aronyu127 commented on db/migrate/20250303055612_add_column_to_users.rb:
```diff
@@ -0,0 +1,8 @@
+  def change
+    add_column :users, :first_name, :string
+    add_column :users, :last_name, :string
+    add_column :users, :job_title, :string
```
db/migrate/20250303091109_add_confirmed_column_to_user.rb
db/migrate/20250303055612_add_column_to_users.rb

這兩個可以砍掉 跟 db/migrate/20250303032628_create_join_table_users_roles.rb 併在一起

Aronyu127 commented on db/migrate/20250303235950_add_default_users.rb:
```diff
@@ -0,0 +1,12 @@
+class AddDefaultUsers < ActiveRecord::Migration[6.1]
+  def change
```
這類初始設定通常會寫在 db/seeds.rb

https://develclan.com/seeding-database-ruby-on-rails/

Aronyu127 commented on test/fixtures/roles.yml:
```diff
@@ -0,0 +1,13 @@
+# Read about fixtures at https://api.rubyonrails.org/classes/ActiveRecord/FixtureSet.html
+
+admin:
```
可試試用 https://github.com/thoughtbot/factory_bot 產生model相關資料

Aronyu127 reviewed (commented):
大致上蠻完整的，一些小建議

kevin5603 committed:
refactor and fix bug base on suggestion
- Replace fixture test data with FactoryBot
- Implement CanCanCan for view authorization
- Ensure role_ids are permitted in strong parameters
- Refactor

kevin5603 committed:
fix bug

Aronyu127 commented on app/models/role.rb:
```diff
@@ -1,3 +1,7 @@
   has_and_belongs_to_many :users
+
+  def self.get_default_role
+    find_or_create_by!(name: 'regular_user')
```
這個應該要加到 Seed  
讓預設的 user 一開始就存在

然後這邊可以用 model scope  or 常數 去定義 default role 是哪一筆

或者是其實 也可以不用給 預設role 
直接在 ability 定義預設可以做的事情就好
role 可以當作是有要多給什麼權限時才賦予的

Aronyu127 reviewed (commented):
🙏 剩一個小地方

Aronyu127 commented on app/abilities/ability.rb:
```diff
@@ -0,0 +1,18 @@
+      can [:read, :update, :destroy], User
+    end
+    if @roles.include?('manager')
+      can :read, User
```
**Regular users can only view and edit their own profiles.**

好像少了這個

kevin5603 committed:
Add regular user can view and edit own profile & refactor

kevin5603 commented:
ad profile
<img width="335" alt="image" src="https://github.com/user-attachments/assets/9e761851-0271-429b-afb1-9f659171c27e" />
edit profile
<img width="586" alt="image" src="https://github.com/user-attachments/assets/5a5c9db8-dd2e-4c4d-935c-c8373838aedd" />

Aronyu127 commented on app/controllers/users_controller.rb:
```diff
@@ -1,21 +1,17 @@
 class UsersController < ApplicationController
   before_action :authenticate_user!
+  before_action :get_roles, only: %i[show edit update]
```
這種就蠻適合 抽成 helper 給前端用的

Aronyu127 commented on app/controllers/users_controller.rb:
```diff
@@ -0,0 +1,34 @@
+  end
+
+  def update
+    if @user.update(user_params)
```
https://github.com/kevin5603/user-management/pull/2/files#r1988394360

應該有一樣的問題
